middleware/rate_limiter.go

```diff
@@ -0,0 +1,219 @@
+)
+
+// DefaultRateLimiterConfig defines default values for RateLimiterConfig
+var DefaultRateLimiterConfig = RateLimiterConfig{
+	Skipper: DefaultSkipper,
+	IdentifierExtractor: func(ctx echo.Context) (string, error) {
+		id := ctx.RealIP()
+		return id, nil
+	},
+	ErrorHandler: func(context echo.Context) error {
+		return context.JSON(http.StatusTooManyRequests, nil)
+	},
+	DenyHandler: func(context echo.Context) error {
+		return context.JSON(http.StatusForbidden, nil)
+	},
+}
+
+/*
+RateLimiter returns a rate limiting middleware
+
+	e := echo.New()
+
+	var inMemoryStore = NewRateLimiterMemoryStore(
+		RateLimiterMemoryStoreConfig{rate: 1, burst: 3, expiresIn: 3 * time.Minute}
+	)
+
+	e.GET("/rate-limited", func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
+	}, RateLimiter(inMemoryStore))
+*/
+func RateLimiter(store RateLimiterStore) echo.MiddlewareFunc {
+	config := DefaultRateLimiterConfig
+	config.Store = store
+
+	return RateLimiterWithConfig(config)
+}
+
+/*
+RateLimiterWithConfig returns a rate limiting middleware
+
+	e := echo.New()
+
+	var inMemoryStore = NewRateLimiterMemoryStore(
+		RateLimiterMemoryStoreConfig{rate: 1, burst: 3, expiresIn: 3 * time.Minute}
+	)
+
+	config := RateLimiterConfig{
+		Skipper: DefaultSkipper,
+		IdentifierExtractor: func(ctx echo.Context) (string, error) {
+			id := ctx.RealIP()
+			return id, nil
+		},
+		ErrorHandler: func(context echo.Context) error {
+			return context.JSON(http.StatusTooManyRequests, nil)
+		},
+		DenyHandler: func(context echo.Context) error {
+			return context.JSON(http.StatusForbidden, nil)
+		},
+	}
+
+	e.GET("/rate-limited", func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
+	}, RateLimiterWithConfig(config))
+*/
+func RateLimiterWithConfig(config RateLimiterConfig) echo.MiddlewareFunc {
+	if config.Skipper == nil {
+		config.Skipper = DefaultRateLimiterConfig.Skipper
+	}
+	if config.IdentifierExtractor == nil {
+		config.IdentifierExtractor = DefaultRateLimiterConfig.IdentifierExtractor
+	}
+	if config.ErrorHandler == nil {
+		config.ErrorHandler = DefaultRateLimiterConfig.ErrorHandler
+	}
+	if config.DenyHandler == nil {
+		config.DenyHandler = DefaultRateLimiterConfig.DenyHandler
+	}
+	if config.Store == nil {
+		panic("Store configuration must be provided")
+	}
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			if config.BeforeFunc != nil {
+				config.BeforeFunc(c)
+			}
+
+			identifier, err := config.IdentifierExtractor(c)
+			if err != nil {
+				return config.DenyHandler(c)
+			}
+
+			if !config.Store.Allow(identifier) {
+				return config.ErrorHandler(c)
+			}
+			return next(c)
+		}
+	}
+}
+
+type (
+	// RateLimiterMemoryStore is the built-in store implementation for RateLimiter
+	RateLimiterMemoryStore struct {
+		visitors    map[string]*Visitor
+		mutex       sync.Mutex
```

Comment:
maybe using `sync.RWMutex` would help. For parts what are reading map read lock `mutex.RLock()` would be used and parts where delete is done `mutex.Lock()` is used.

Reply:
This is why I wanted to have benchmarks for the store in this PR now. This will allow to improve this later on.
The sharded implementation helps to reduce locks and helps for high concurrency (the test uses 100 concurrent calls for example). But this should be done in a different PR then to get this one merged I think.

Reply:
I more thing - ratelimiter itself has [internal locking](https://github.com/golang/time/blob/7e3f01d253248a0a5694eb5b7376dfea18b6397e/rate/rate.go#L56) mechanism so extreme burst from one identifier would cause contention on that limiter. 

anyway. there are examples for mutex benchmarks
*  https://stackoverflow.com/questions/57562606/why-does-sync-mutex-largely-drop-performance-when-goroutine-contention-is-more-t 
* https://github.com/golang/time/blob/7e3f01d253248a0a5694eb5b7376dfea18b6397e/rate/rate_test.go#L457

and https://golang.org/doc/diagnostics.html profiler has profile for mutex contentions
> mutex: Mutex profile reports the lock contentions. When you think your CPU is not fully utilized due to a mutex contention, use this profile. Mutex profile is not enabled by default, see runtime.SetMutexProfileFraction to enable it.

without real facts I still believe that for in-memory/single instance application mutex contention is not a real problem.  your app would die anyway on business logic (db access) - assuming contention is due different identifiers. and for same identifier limiter own implementation has lock.

Reply:
Yes, agreed. My intention was to ensure we have no races and optimize later if required, so the benchmarks were already used for that. My experiments with the sharded implementation were out of personal interest in that topic.

Thanks for the links very interesting. The benchmark tests should  also cover these extremes probably to detect issues when modifying our default implementation. But that will also be left to another PR.
But thanks for the inputs here, very valuable @aldas 

Just to state a view findings:

* Concurrency above 100 parallel requests already causes measurable lock contention caused by the mutex (sharding helps here of course) 
* High number of different IP addresses (above 16000) will cause increased issues with Go's map implementation 
* Burst 0 is not allowed, as it will reject all events and does not make sense in our usecase

Reply:
If there would be disclaimer in code and docs and links to different implementations - simpler implementation would be good enough. 

Brief googling gives me:
* https://github.com/didip/tollbooth_echo
* https://github.com/ulule/limiter-examples/blob/master/echo/main.go

a lot of cases can even be covered with extremely simplistic/limited approach as 
```go
func ConcurrentRequestLimit(concurrentRequestLimit int) echo.MiddlewareFunc {
	sem := make(chan struct{}, concurrentRequestLimit)
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			sem <- struct{}{}
			defer func() {
				<-sem
			}()

			return next(c)
		}
	}
}
```

Reply:
Limiting concurrency is very different then rate limiting. This PR is about having a simple implementation available in echo core that will work for the standard use case and should be easy to use. Do you see the current proposal to be too complex to use?

We are down to this currently:

```go
    e := echo.New()
    limiterStore := middleware.NewRateLimiterMemoryStore(20)
    e.Use(middleware.RateLimiter(limiterStore))
```

Disclaimers in the code could link to echo-contrib where other (also 3rdparty) rate limiting implementations are listed. This would be an awesome-echo list more or less.

Reply:
Seems nicely clean. Good thing that burst is not included in shortest way of initialization. I'm not most brightest person and I had to search around what it means in rate limiting sense and have it explained like I'm 5 old - how it relates to rate, can it be higher/lower? etc.

Reply:
I guess we could add a very short explanation in the Config struct and point to the x/time/rate docs.
I need to add a PR in echox for the docs anyway where we can add a section Although to be honest in real world testing the results were not what I expected, so I guess I'm also a 5 year old boy ;-)